src/NServiceBus.Core.Tests/Recoverability/RecoverabilityBehaviorTests.cs

```diff
@@ -0,0 +1,144 @@
+        {
+            var dispatcher = new FakeDispatcher();
+            var context = CreateContext();
+            var behavior = CreateBehavior(dispatcher, transactionsEnabled: true);
+
+            await behavior.Invoke(context, () => { throw new Exception(); });
+
+            await behavior.Invoke(context, () => Task.FromResult(0));
+
+            Assert.IsNotNull(dispatcher.TransportOperation.Message);
+        }
+
+        [Test]
+        public async Task ShouldNotCallRestOfThePipelineWhenMovingMesasgeToErrorQueue_WhenRunningWithTransactions()
+        {
+            var dispatcher = new FakeDispatcher();
+            var context = CreateContext();
+            var behavior = CreateBehavior(dispatcher, transactionsEnabled: true);
+            var pipelineCalled = false;
+
+            await behavior.Invoke(context, () => { throw new Exception(); });
+
+            await behavior.Invoke(context, () => {
+                pipelineCalled = true;
+                return Task.FromResult(0);
+            });
+
+            Assert.IsFalse(pipelineCalled);
+        }
+
+        [Test]
+        public async Task ShouldSkipRetryForDeserializationErrors()
+        {
+            var dispatcher = new FakeDispatcher();
+            var context = CreateContext();
+            var behavior = CreateBehavior(dispatcher, flrEnabled: true, slrEnabled: true, transactionsEnabled: true);
+
+            await behavior.Invoke(context, () => { throw new MessageDeserializationException(string.Empty); });
+
+            await behavior.Invoke(context, () => Task.FromResult(0));
+
+            Assert.IsNotNull(dispatcher.TransportOperation);
+            Assert.AreEqual(ErrorQueueAddress, dispatcher.TransportOperation.Destination);
+        }
+
+        RecoverabilityBehavior CreateBehavior(IDispatchMessages messageDispatcher, bool flrEnabled = false, bool slrEnabled = false, bool transactionsEnabled = false)
+        {
+            var failureStorage = new FailureInfoStorage(1000);
+
+            var flrHandler = flrEnabled ? new FirstLevelRetriesHandler(failureStorage, new FirstLevelRetryPolicy(2)) : null;
+
+            var slrHandler = slrEnabled 
+                ? new SecondLevelRetriesHandler(new DefaultSecondLevelRetryPolicy(2, TimeSpan.FromSeconds(3)), failureStorage, new DelayedRetryExecutor("input", messageDispatcher))
+                : null; 
+
+            var errorBehavior = new MoveFaultsToErrorQueueHandler(
+                new FakeCriticalError(),
+                new FailureInfoStorage(10),
+                new MoveToErrorsActionExecutor(messageDispatcher, ErrorQueueAddress, new Dictionary<string, string>()));
+
+            return new RecoverabilityBehavior(flrHandler, slrHandler, errorBehavior, flrEnabled, slrEnabled, transactionsEnabled);
+        }
+
+        static TestableTransportReceiveContext CreateContext(string messageId = "message-id", FakeEventAggregator eventAggregator = null)
```

Comment:
I think you never pass a message id to CreateContext in the tests, so you simplify this a bit? (or change the tests which demonstrate transaction rollbacks to use different context instances with the same message id)